shell/platform/darwin/macos/framework/Source/FlutterView.mm

```diff
@@ -3,31 +3,250 @@
-    self.wantsBestResolutionOpenGLSurface = YES;
+
+    [self setWantsLayer:YES];
+
+    // covers entire view
+    contentLayer = [[CALayer alloc] init];
+    [self.layer addSublayer:contentLayer];
+
+    synchronizer = fml::MakeRefCounted<ResizeSynchronizer>(self);
+
+    [self.openGLContext makeCurrentContext];
+
+    glGenFramebuffers(1, &_frameBufferId);
+    glBindFramebuffer(GL_FRAMEBUFFER, _frameBufferId);
+
+    glGenTextures(1, &_backingTexture);
+    glBindTexture(GL_TEXTURE_RECTANGLE_ARB, _backingTexture);
+    glTexParameterf(GL_TEXTURE_RECTANGLE_ARB, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+    glTexParameterf(GL_TEXTURE_RECTANGLE_ARB, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+    glBindTexture(GL_TEXTURE_RECTANGLE_ARB, 0);
+
+    _delegate = delegate;
   }
   return self;
 }
 
+- (void)recreateSurfaceWithScaledSize:(CGSize)scaledSize {
+  if (_ioSurface) {
+    CFRelease(_ioSurface);
+  }
+
+  [self.openGLContext makeCurrentContext];
+
+  unsigned pixelFormat = 'BGRA';
+  unsigned bytesPerElement = 4;
+
+  size_t bytesPerRow =
+      IOSurfaceAlignProperty(kIOSurfaceBytesPerRow, scaledSize.width * bytesPerElement);
+  size_t totalBytes = IOSurfaceAlignProperty(kIOSurfaceAllocSize, scaledSize.height * bytesPerRow);
+  NSDictionary* options = @{
+    (id)kIOSurfaceWidth : @(scaledSize.width),
+    (id)kIOSurfaceHeight : @(scaledSize.height),
+    (id)kIOSurfacePixelFormat : @(pixelFormat),
+    (id)kIOSurfaceBytesPerElement : @(bytesPerElement),
+    (id)kIOSurfaceBytesPerRow : @(bytesPerRow),
+    (id)kIOSurfaceAllocSize : @(totalBytes),
+  };
+  _ioSurface = IOSurfaceCreate((CFDictionaryRef)options);
+
+  glBindTexture(GL_TEXTURE_RECTANGLE_ARB, _backingTexture);
+
+  CGLTexImageIOSurface2D(CGLGetCurrentContext(), GL_TEXTURE_RECTANGLE_ARB, GL_RGBA,
+                         int(scaledSize.width), int(scaledSize.height), GL_BGRA,
+                         GL_UNSIGNED_INT_8_8_8_8_REV, _ioSurface, 0 /* plane */);
+  glBindTexture(GL_TEXTURE_RECTANGLE_ARB, 0);
+
+  glBindFramebuffer(GL_FRAMEBUFFER, _frameBufferId);
+  glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_RECTANGLE_ARB,
+                         _backingTexture, 0);
+}
+
+- (void)reshaped {
+  if (active) {
+    CGSize scaledSize = [self convertRectToBacking:self.bounds].size;
+    synchronizer->BeginResize(scaledSize, ^{
+      [_delegate viewDidReshape:self];
+    });
+  }
+}
+
 #pragma mark - NSView overrides
 
+- (void)setFrameSize:(NSSize)newSize {
+  [super setFrameSize:newSize];
+  [self reshaped];
+}
+
+- (void)commit {
+  [self.openGLContext flushBuffer];
+
+  [CATransaction begin];
+  [CATransaction setDisableActions:YES];
+  self.layer.frame = self.bounds;
+  self.layer.sublayerTransform = CATransform3DTranslate(CATransform3DMakeScale(1, -1, 1), 0,
+                                                        -self.layer.bounds.size.height, 0);
+  contentLayer.frame = self.layer.bounds;
+
+  [contentLayer setContents:nil];
```

Comment:
add a comment explaining why we need to set to `nil` before setting it to the contents of the `ioSurface`